tools/json_tools/generic_guns_validator.py

```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python3
+
+import os
+import sys
+import util
+
+SCRIPT_DIR = os.path.abspath(os.path.dirname(__file__))
+GG_DIR = os.path.normpath(os.path.join(
+    SCRIPT_DIR, '..', '..', 'data', 'mods', 'Generic_Guns'))
+
+# We might want to also warn about grenades and rockets, but leaving them
+# whitelisted for now.
+AMMO_TYPE_WHITELIST = {
+    '40x46mm', # Grenade
+    'atgm', # Rocket
+    'barb',
+    'battery',
+    'BB',
+    'blunderbuss',
+    'bolt',
+    'charcoal',
+    'chemical_spray',
+    'fishspear',
+    'flammable',
+    'm235', # Rocket
+    'metal_rail',
+    'nail',
+    'paintball',
+    'pebble',
+    'plasma',
+    'rock',
+    'signal_flare',
+    'weldgas',
+}
+SKILL_WHITELIST = {
+    'archery',
+    'launcher',
+}
+# Whitelisted specific ids are mostly obsolete stuff.
+# This could go away if obsolete stuff were explicitly marked as such.
+ID_WHITELIST = {
+    # Guns
+    'coilgun',
+    'ftk93',
+    'l_bak_223',
+    'pneumatic_shotgun',
+    'rifle_223',
+    'rifle_3006',
+    'rifle_308',
+    'surv_carbine_223',
+    # Magazines
+    '223_speedloader5',
+}
+
+def items_of_type(data, type):
+    result = []
+    for i in data:
+        if i['type'] == type:
+            result.append(i)
+    return result
+
+def get_ids(items):
+    result = set()
+    for i in items:
+        id = i.get('id', [])
+        if isinstance(id, str):
+            result.add(id)
+        else:
+            result.update(id)
+    return result
+
+def get_id(i):
+    if 'id' in i:
+        return i['id']
+    if 'abstract' in i:
+        return i['abstract']
+    raise RuntimeError('item lacks id: %r' % i)
+
+def items_for_which_any_ancestor(items, pred):
+    items_by_id = {get_id(i): i for i in items}
+    result = []
+    for i in items:
+        this_i = i
+        matches = False
+        while True:
```

Comment:
What if `copy-from` will make a cycle? Is it safe to do an endless loop here and in `items_for_which_all_ancestors`?

Reply:
I think a `copy-from` cycle would be caught as an error by the game, but you're right it's probably best to catch it here too.

Reply:
Tweaked the code to avoid this concern.  And made it clearer at the same time, so thanks for the suggestion.